src/mod_ping.erl

```diff
@@ -1,130 +1,101 @@
-                timeout_action = none,
-                ping_req_timeout = ?DEFAULT_PING_REQ_TIMEOUT,
-                timers = dict:new()}).
-
--type state() :: #state{}.
-
 %%====================================================================
-%% API
+%% Info Handler
 %%====================================================================
-start_link(Host, Opts) ->
-    Proc = gen_mod:get_module_proc(Host, ?MODULE),
-    gen_server:start_link({local, Proc}, ?MODULE, [Host, Opts], []).
 
-start_ping(Host, JID) when JID#jid.lresource =/= <<>> ->
-    %% Guard check above ensures we are not adding a user without a
-    %% resource by accident. If this happens the server ends up in a
-    %% hot loop of iq-error replies.
-    Proc = gen_mod:get_module_proc(Host, ?MODULE),
-    gen_server:cast(Proc, {start_ping, JID});
-start_ping(_Host, _JID) -> ok.
+handle_info(init, HandlerState, C2SState) ->
+    {add_ping_timer(HandlerState, C2SState), C2SState};
+handle_info(send_ping, HandlerState, C2SState) ->
+    JID = ejabberd_c2s_state:jid(C2SState),
+    Server = ejabberd_c2s_state:server(C2SState),
+    route_ping_iq(JID, Server),
+    {add_ping_timer(HandlerState, C2SState), C2SState};
+handle_info(timeout, HandlerState, C2SState) ->
+    JID = ejabberd_c2s_state:jid(C2SState),
+    Server = ejabberd_c2s_state:server(C2SState),
+    ejabberd_hooks:run(user_ping_timeout, Server, [JID]),
+    case gen_mod:get_module_opt(Server, ?MODULE, timeout_action, none) of
+        kill -> ejabberd_c2s:stop(self());
+        _ -> ok
+    end,
+    {HandlerState, C2SState}.
 
-stop_ping(Host, JID) ->
-    Proc = gen_mod:get_module_proc(Host, ?MODULE),
-    gen_server:cast(Proc, {stop_ping, JID}).
+add_ping_timer(HandlerState, C2SState) ->
+    cancel_timer(HandlerState),
+    Server = ejabberd_c2s_state:server(C2SState),
+    PingInterval = gen_mod:get_module_opt(Server, ?MODULE, ping_interval, ?DEFAULT_PING_INTERVAL),
+    erlang:send_after(PingInterval * 1000, self(), {mod_ping, send_ping}).
 
-%%====================================================================
-%% gen_mod callbacks
-%%====================================================================
-start(Host, Opts) ->
-    Proc = gen_mod:get_module_proc(Host, ?MODULE),
-    PingSpec = {Proc, {?MODULE, start_link, [Host, Opts]},
-                transient, 2000, worker, [?MODULE]},
-    supervisor:start_child(?SUPERVISOR, PingSpec).
+route_ping_iq(JID, Server) ->
+    PingReqTimeout = gen_mod:get_module_opt(Server, ?MODULE, ping_req_timeout,
+                                            ?DEFAULT_PING_REQ_TIMEOUT),
+    IQ = #iq{type = get,
+             sub_el = [#xmlel{name = <<"ping">>,
+                              attrs = [{<<"xmlns">>, ?NS_PING}]}]},
+    Pid = self(),
+    F = fun(_, _, _, timeout) ->
+               Pid ! {mod_ping, timeout};
+           (_From, _To, Acc, _Response) ->
+                Acc
+        end,
+    From = jid:make(<<"">>, Server, <<"">>),
```

Comment:
I think it's safe to use
```suggestion
    From = jid:make_noprep(<<"">>, Server, <<"">>),
```

since the value in `Server` is already stringprepped.